File: 03_binary_search/01_search_a_2d_matrix/01_search_a_2d_matrix.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def mean(a, b): return int((a + b) / 2)
        begin = 0
        end = len(matrix)
        mid = mean(begin, end)
        while begin != mid:
            if matrix[mid][0] == target:
                return True
            elif matrix[mid][0] > target:
                end = mid
                mid = mean(begin, end)
            else:  # matrix[mid][0] < target
                begin = mid
                mid = mean(begin, end)

        nums: List[int] = matrix[mid]
        begin = 0
        end = len(nums)
        mid = mean(begin, end)
        while begin != mid:
            if nums[mid] == target:
                return True
            elif nums[mid] > target:
                end = mid
                mid = mean(begin, end)
            else:  # nums[mid] < target
                begin = min(mid + 1, end - 1)
                mid = mean(begin, end)

        return nums[mid] == target


if __name__ == "__main__":
    solution = Solution()
    matrix = [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]]
    target = 13
    print(solution.searchMatrix(matrix, target))

File: 03_binary_search/01_search_a_2d_matrix/01_search_a_2d_matrix.cpp
// Standard: C++11

#include <vector>
#include <cstdint>

using std::vector;

class Solution
{
  public:
    bool searchMatrix(vector<vector<int>>& matrix, int target)
    {
        auto   mean = [](int a, int b) { return (a + b) / 2; };
        size_t begin = 0;
        size_t end = matrix.size();
        size_t mid = mean(begin, end);
        while (begin != mid)
        {
            if (matrix[mid][0] == target)
                return true;
            else if (matrix[mid][0] > target)
            {
                end = mid;
                mid = mean(begin, end);
            }
            else // matrix[mid][0] < target
            {
                begin = std::min(mid, end - 1);
                mid = mean(begin, end);
            }
        }

        const auto& vec = matrix[mid];
        begin = 0;
        end = vec.size();
        mid = mean(begin, end);
        while (begin != mid)
        {
            if (vec[mid] == target)
                return true;
            else if (vec[mid] > target)
            {
                end = mid;
                mid = mean(begin, end);
            }
            else // vec[mid] < target
            {
                begin = std::min(mid + 1, end - 1);
                mid = mean(begin, end);
            }
        }
        return vec[mid] == target;
    }
};

#include <iostream>

int main()
{
    vector<vector<int>> matrtix = {
        {1, 3, 5, 7}, {10, 11, 16, 20}, {23, 30, 34, 60}};
    int      target = 13;
    Solution solution;
    std::cout << target << " "
              << ((solution.searchMatrix(matrtix, target)) ? "is " : "is not ")
              << "in matrix.\n";
}
